### PR TITLE
Authentication rework [depends on #2072]

### PR DESCRIPTION
**Pull request by thepieterdc**
This pull request reworks the inner authentication mechanisms to decouple institutions from login providers.

- [x] Providers
  - [x] SAML
  - [x] SmartSchool
  - [x] Google OAuth2
  - [x] Office365
- [x] Identities
- [x] Migrate all existing institutions to providers
  - [x] SAML
  - [x] SmartSchool
  - [x] Google OAuth2
  - [x] Office365
- [x] Creation of new institutions
- [x] Migrate all existing users to identities
- [x] Update fallback behaviour for usernames
- [x] UserController#new
- [x] Tests

Future work in https://github.com/dodona-edu/dodona/issues/2067

### CODE REVIEW

File: db/migrate/20200620113027_extract_usernames_to_identities.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class ExtractUsernamesToIdentities < ActiveRecord::Migration[6.0]
+  def change
+    User.find_each do |user|
+      if user.institution.present?
+        Identity.create provider: user.institution.providers.first,
+                        identifier: user.username,
```
Comment (bmesuere) [inline db/migrate/20200620113027_extract_usernames_to_identities.rb:6]: I assume this is just a temp value that will be overwritten by a provider specific identifier when the user first signs in? Or do we already store the uid as username right now?

Comment (thepieterdc) [reply db/migrate/20200620113027_extract_usernames_to_identities.rb:6]: > I assume this is just a temp value that will be overwritten by a provider specific identifier when the user first signs in? Or do we already store the uid as username right now?

The uid is actually already stored in the username field.

Comment (thepieterdc) [reply db/migrate/20200620113027_extract_usernames_to_identities.rb:6]: https://github.com/dodona-edu/dodona/blob/7e993008927284cead4441402b0bd1c9c779a13e/app/models/user.rb#L261

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -46,9 +46,12 @@ def after_sign_out_path_for(_resource)
   end
 
   Warden::Manager.after_authentication do |user, auth, _opts|
+    # SAML authentication.
     if user.institution.nil?
       idp = Institution.find_by(short_name: auth.env['rack.session'][:current_idp])
-      user.update(institution: idp)
+      user.institution = idp
+      user.identities.build(provider: idp.providers.first, identifier: user.username)
```
Comment (charvp) [inline app/controllers/application_controller.rb:53]: Why is `idp.providers.first` used here? In this phase each institution only has one provider, I guess, but don't we want to find the correct provider using the same data as we use to find the institution here?

Comment (charvp) [inline app/controllers/application_controller.rb:53]: Right now the user's email address is used to find users. Is switching to `username` something that will work for all (SAML) institutions?

Comment (thepieterdc) [reply app/controllers/application_controller.rb:53]: > Right now the user's email address is used to find users. Is switching to `username` something that will work for all (SAML) institutions?

This is not true: https://github.com/dodona-edu/dodona/blob/da323f58c9f35ecc82969d544a6a367f869a6817/lib/SAML/my_resource_validator.rb#L13

Comment (thepieterdc) [reply app/controllers/application_controller.rb:53]: I have added this to the list of todo's in #2067

**Pull request merged**